File: RMV.ML.MNIST.CS/MultiLayerNet.cs
using MathNet.Numerics.Distributions;
using MathNet.Numerics.LinearAlgebra;

using RMV.ML.Network.Common;

namespace RMV.ML.MNIST.CS;

/// <summary>
/// Fully connected deep neural network with arbitrary hidden layers.
/// </summary>
public class MultiLayerNet
{
	readonly int inputSize, outputSize; // input and output dimensions
	readonly List<int> HiddenSizeList;  // list of hidden layer sizes
	readonly int hiddenLayerNum;        // number of hidden layers
	readonly double weightDecayLambda;  // weight decay coefficient

	/// <summary>
	/// Weight matrices indexed by layer 
	/// </summary>
	Matrix<double>[] Weights { get; }

	/// <summary>
	/// Bias vectors indexed by layer 
	/// </summary>
	Vector<double>[] Biases { get; }

	/// <summary>
	/// Ordered list of all layers (affine + activation) for forward/backward traversal.
	/// </summary>
	readonly List<ILayer> Layers = [];

	/// <summary>
	/// Affine layers indexed by layer number for gradient extraction.
	/// </summary>
	readonly Affine[] AffineLayers;

	SoftmaxWithLoss LastLayer { get; } = new();

	readonly IOptimizer optimizer;

	/// <summary>
	/// Construct a multi-layer fully connected neural network based on the provided settings.
	/// </summary>
	public MultiLayerNet( AppSettings settings )
	{
		this.inputSize = settings.Input;
		this.HiddenSizeList = [ .. settings.Hidden ];
		this.outputSize = settings.Output;
		this.hiddenLayerNum = settings.Hidden.Length;
		this.weightDecayLambda = settings.Decay;

		int totalLayers = hiddenLayerNum + 1;
		this.Weights = new Matrix<double>[ totalLayers ];
		this.Biases = new Vector<double>[ totalLayers ];
		this.AffineLayers = new Affine[ totalLayers ];

		this.optimizer = BuildOptimizer( settings );

		InitWeight( settings.Activation );

		for( int i = 0; i < hiddenLayerNum; i++ )
		{
			var affine = new Affine( this.Weights[ i ], this.Biases[ i ] );
			this.AffineLayers[ i ] = affine;
			this.Layers.Add( affine );

			if( settings.Activation == ActivationType.Relu )
				this.Layers.Add( new Relu() );
			else if( settings.Activation == ActivationType.Sigmoid )
				this.Layers.Add( new Sigmoid() );
		}

		int lastIdx = hiddenLayerNum;
		var lastAffine = new Affine( this.Weights[ lastIdx ], this.Biases[ lastIdx ] );
		this.AffineLayers[ lastIdx ] = lastAffine;
		this.Layers.Add( lastAffine );
	}


	/// <summary>
	/// Creates Optimizer based on configuration settings 
	/// </summary>	
	static IOptimizer BuildOptimizer( AppSettings settings )
	{
		return settings.Optimizer switch {
			Optimizer.SGD => new SGD(),
			Optimizer.Momentum => new Momentum(),
			Optimizer.Adam => new Adam(),
			Optimizer.Nesterov => new Nesterov(),
			Optimizer.AdaGrad => new AdaGrad(),
			Optimizer.RmsProp => new RmsProp(),
			_ => throw new Exception( $"Unsupported optimizer: {settings.Optimizer}" )
		};
	}

	/// <summary>
	/// Initialize weights based on the specified activation function using appropriate initialization methods
	/// </summary>	
	void InitWeight( ActivationType activation )
	{
		var allSizeList = new List<int> { inputSize };
		allSizeList.AddRange( HiddenSizeList );
		allSizeList.Add( outputSize );

		for( int i = 0; i < allSizeList.Count - 1; i++ )
		{
			double scale = 0.01;
			if( activation == ActivationType.Relu )
				scale = Math.Sqrt( 2.0 / allSizeList[ i ] ); // ReLU (He initialization)
			else if( activation == ActivationType.Sigmoid )
				scale = Math.Sqrt( 1.0 / allSizeList[ i ] ); // Sigmoid (Xavier initialization)
			//else if( double.TryParse( weightInitStd, out double customScale ) )
			//	scale = customScale;

			this.Weights[ i ] = Matrix<double>.Build.Random( allSizeList[ i ], allSizeList[ i + 1 ], new Normal( 0.0, scale ) );
			this.Biases[ i ] = Vector<double>.Build.Dense( allSizeList[ i + 1 ], 0.0 );
		}
	}

	/// <summary>
	/// Perform a forward pass through the network to compute the output predictions for the given input data.
	/// </summary>
	/// <param name="x">Input data matrix</param>
	/// <returns>Output predictions matrix</returns>
	Matrix<double> Predict( Matrix<double> x )
	{
		this.Layers.ForEach( l => x = l.Forward( x ) );

		return x;
	}

	/// <summary>
	/// Compute the loss (cost) of the network's predictions compared to the true labels, including weight decay regularization if specified.
	/// </summary>
	/// <param name="x">Input data matrix</param>
	/// <param name="t">True labels matrix</param>
	/// <returns>Loss value</returns>
	public double Loss( Matrix<double> x, Matrix<double> t )
	{
		var y = Predict( x );

		double weightDecay = 0.0;

		for( int i = 0; i < Weights.Length; i++ )
		{
			double sumSqr = this.Weights[ i ].Enumerate().Sum( val => val * val );

			weightDecay += 0.5 * weightDecayLambda * sumSqr;
		}

		return LastLayer.Forward( y, t ) + weightDecay;
	}

	/// <summary>
	/// Calculate the accuracy of the network's predictions by comparing the predicted class labels to the true labels
	/// </summary>
	/// <param name="x">Input data matrix</param>
	/// <param name="t">True labels matrix</param>
	/// <returns>Accuracy value</returns>
	public (double, List<int>, List<int>) Accuracy( Matrix<double> x, Matrix<double> t )
	{
		var y = Predict( x );
		int batchSize = y.RowCount;
		int correct = 0;
		List<int> errors = [];
		List<int> indexes = [];

		for( int i = 0; i < batchSize; i++ )
		{
			int yPred = y.Row( i ).MaximumIndex();
			int tIndex = t.ColumnCount == 1 ? ( int )t[ i, 0 ] : t.Row( i ).MaximumIndex();

			if( yPred == tIndex ) correct++;  // count correct predictions
			else // store misclassified sample information 
			{
				errors.Add( yPred );  // store predicted label 
				indexes.Add( i );     // store image index 
			}		
		}

		return (( double )correct / batchSize, errors, indexes);
	}

	/// <summary>
	/// Update the network's parameters using the computed gradients and the optimizer.
	/// </summary>
	/// <param name="x">Input data matrix</param>
	/// <param name="t">True labels matrix</param>
	public void Update( Matrix<double> x, Matrix<double> t )
	{
		var (dW, dB) = Gradient( x, t ); // compute gradients via backprop

		this.optimizer.Update( this.Weights, this.Biases, dW, dB ); // update parameters directly
	}

	/// <summary>
	/// Run backpropagation to compute the gradients of the loss with respect to all weights and biases in the network,
	/// including weight decay regularization for the weight gradients.
	/// </summary>
	/// <param name="x">Input data matrix</param>
	/// <param name="t">True labels matrix</param>
	/// <returns>Tuple of weight gradients and bias gradients arrays (0-based indexing)</returns>
	(Matrix<double>[], Vector<double>[]) Gradient( Matrix<double> x, Matrix<double> t )
	{
		Loss( x, t ); // Forward pass

		Matrix<double> dout = LastLayer.Backward( 1.0 ); // Backward pass

		for( int i = Layers.Count - 1; i >= 0; i-- )
		{
			dout = Layers[ i ].Backward( dout );
		}

		var dW = new Matrix<double>[ AffineLayers.Length ];
		var dB = new Vector<double>[ AffineLayers.Length ];

		for( int i = 0; i < AffineLayers.Length; i++ )
		{
			dW[ i ] = AffineLayers[ i ].dW + this.Weights[ i ].Multiply( weightDecayLambda ); // W_grad = dW + lambda * W
			dB[ i ] = AffineLayers[ i ].dB;
		}

		return (dW, dB);
	}

}

File: RMV.ML.MNIST.PY/Multi_Layer_Net.py
import jax  # type: ignore
import jax.numpy as jnp  # type: ignore
from jax import random  # type: ignore
from dataclasses import dataclass, field
from enum import Enum
from typing import Union

from layers import Relu, Sigmoid, Affine, SoftmaxWithLoss
from optimizer import SGD, Momentum, Nesterov, Adam, AdaGrad, RmsProp


# ---------------------------------------------------------------------------
# Enums
# ---------------------------------------------------------------------------
class ActivationType(Enum):
    Relu    = "Relu"
    Sigmoid = "Sigmoid"


class OptimizerType(Enum):
    SGD      = "SGD"
    Momentum = "Momentum"
    Nesterov = "Nesterov"
    Adam     = "Adam"
    AdaGrad  = "AdaGrad"
    RmsProp  = "RmsProp"

# ---------------------------------------------------------------------------
# AppSettings
# ---------------------------------------------------------------------------
@dataclass
class AppSettings:
    input:       int              = 784
    hidden:      list[int]        = field(default_factory=lambda: [100])
    output:      int              = 10
    activation:  ActivationType   = ActivationType.Relu
    optimizer:   OptimizerType    = OptimizerType.SGD
    decay:       float            = 0.0
    rate:        float            = 0.01
    momentum:    float            = 0.9
    iterations:  int              = 10000
    batch:       int              = 100
    epoch:       int              = 10
    weight_init_std: float        = 0.01
    print_interval: int          = 100
    stagnation: int              = 100
    train_path: str              = "D:\\MNIST\\mnist_train.csv"
    test_path:  str              = "D:\\MNIST\\mnist_test.csv"
    error_path: str              = "D:\\MNIST\\errors.txt"
    index_path: str              = "D:\\MNIST\\indexes.txt"


# ---------------------------------------------------------------------------
# MultiLayerNet
# ---------------------------------------------------------------------------
class MultiLayerNet:
    """
    Fully connected deep neural network with arbitrary hidden layers.
    Mirrors MultiLayerNet.cs — converted to JAX (immutable arrays).
    """

    def __init__(self, settings: AppSettings, key: jax.Array = random.PRNGKey(0)):
        self.input_size        = settings.input
        self.hidden_size_list  = list(settings.hidden)
        self.output_size       = settings.output
        self.hidden_layer_num  = len(settings.hidden)
        self.weight_decay_lambda = settings.decay

        # total_layers = self.hidden_layer_num + 1

        # Mutable lists — reassigned after every optimizer step
        self.weights: list[jnp.ndarray] = []
        self.biases:  list[jnp.ndarray] = []

        # Affine layer objects (hold .dW / .dB after backward)
        self.affine_layers: list[Affine] = []

        # Ordered forward/backward traversal sequence
        self.layers: list[Union[Affine, Relu, Sigmoid]] = []

        self.last_layer = SoftmaxWithLoss()
        self.optimizer  = self._build_optimizer(settings)

        # ---- weight initialisation ----------------------------------------
        key = self._init_weights(settings.activation, settings.weight_init_std, key)

        # ---- build layer stack ---------------------------------------------
        for i in range(self.hidden_layer_num):
            affine = Affine(self.weights[i], self.biases[i])
            self.affine_layers.append(affine)
            self.layers.append(affine)

            if settings.activation == ActivationType.Relu:
                self.layers.append(Relu())
            elif settings.activation == ActivationType.Sigmoid:
                self.layers.append(Sigmoid())

        last_idx   = self.hidden_layer_num
        last_affine = Affine(self.weights[last_idx], self.biases[last_idx])
        self.affine_layers.append(last_affine)
        self.layers.append(last_affine)

    # ------------------------------------------------------------------
    # Construction helpers
    # ------------------------------------------------------------------
    @staticmethod
    def _build_optimizer(settings: AppSettings):
        match settings.optimizer:
            case OptimizerType.SGD:
                return SGD(rate=settings.rate)
            case OptimizerType.Momentum:
                return Momentum(rate=settings.rate, momentum=settings.momentum)
            case OptimizerType.Nesterov:
                return Nesterov(rate=settings.rate, momentum=settings.momentum)
            case OptimizerType.Adam:
                return Adam(rate=settings.rate)
            case OptimizerType.AdaGrad:
                return AdaGrad(rate=settings.rate)
            case OptimizerType.RmsProp:
                return RmsProp(rate=settings.rate)
            case _:
                raise ValueError(f"Unsupported optimizer: {settings.optimizer}")

    def _init_weights(
        self,
        activation: ActivationType,
        weight_init_std: float,
        key: jax.Array,
    ) -> jax.Array:
        """
        He   initialisation for ReLU  : scale = sqrt(2 / fan_in)
        Xavier initialisation for Sigmoid: scale = sqrt(1 / fan_in)
        Default                          : scale = weight_init_std (0.01)
        """
        all_sizes = [self.input_size] + self.hidden_size_list + [self.output_size]

        for i in range(len(all_sizes) - 1):
            fan_in = all_sizes[i]
            fan_out = all_sizes[i + 1]

            if activation == ActivationType.Relu:
                scale = jnp.sqrt(2.0 / fan_in)          # He
            elif activation == ActivationType.Sigmoid:
                scale = jnp.sqrt(1.0 / fan_in)          # Xavier
            else:
                scale = weight_init_std

            key, subkey = random.split(key)
            W = random.normal(subkey, shape=(fan_in, fan_out)) * scale
            b = jnp.zeros(fan_out)

            self.weights.append(W)
            self.biases.append(b)

        return key

    # ------------------------------------------------------------------
    # Forward / Predict
    # ------------------------------------------------------------------
    def _predict(self, x: jnp.ndarray) -> jnp.ndarray:
        """Forward pass through all layers."""
        for layer in self.layers:
            x = layer.forward(x)
        return x

    # ------------------------------------------------------------------
    # Loss
    # ------------------------------------------------------------------
    def loss(self, x: jnp.ndarray, t: jnp.ndarray) -> float:
        """Cross-entropy loss + L2 weight decay."""
        y = self._predict(x)

        weight_decay = sum(
            0.5 * self.weight_decay_lambda * float(jnp.sum(w ** 2))
            for w in self.weights
        )

        return self.last_layer.forward(y, t) + weight_decay

    # ------------------------------------------------------------------
    # Accuracy
    # ------------------------------------------------------------------
    def accuracy( self, x: jnp.ndarray, t: jnp.ndarray ) -> tuple[float, list[int], list[int]]:
        """
        Returns (accuracy, misclassified_predicted_labels, misclassified_indexes).
        t may be one-hot (batch, classes) or label indices (batch, 1).
        """
        y = self._predict(x)
        batch_size = y.shape[0]

        y_pred = jnp.argmax(y, axis=1)  # (batch,)

        if t.shape[1] == 1:             # label indices
            t_idx = t[:, 0].astype(jnp.int32)
        else:                           # one-hot
            t_idx = jnp.argmax(t, axis=1)

        correct = int(jnp.sum(y_pred == t_idx))

        errors:  list[int] = []
        indexes: list[int] = []
        for i in range(batch_size):
            if int(y_pred[i]) != int(t_idx[i]):
                errors.append(int(y_pred[i]))
                indexes.append(i)

        return float(correct / batch_size), errors, indexes

    # ------------------------------------------------------------------
    # Gradient  (backpropagation)
    # ------------------------------------------------------------------
    def _gradient( self, x: jnp.ndarray, t: jnp.ndarray ) -> tuple[list[jnp.ndarray], list[jnp.ndarray]]:
        """Backprop — returns (dW_list, dB_list)."""
        self.loss(x, t)                             # forward pass (populates layer state)

        dout = self.last_layer.backward(1.0)        # backward from loss

        for layer in reversed(self.layers):
            dout = layer.backward(dout)

        dW_list: list[jnp.ndarray] = []
        dB_list: list[jnp.ndarray] = []

        for i, affine in enumerate(self.affine_layers):
            assert affine.dW is not None and affine.dB is not None, "Affine layer has no gradient — was backward() called?"

            # L2 regularisation: grad_W += lambda * W
            dW_list.append(affine.dW + self.weight_decay_lambda * self.weights[i])
            dB_list.append(affine.dB)

        return dW_list, dB_list

    # ------------------------------------------------------------------
    # Update (one training step)
    # ------------------------------------------------------------------
    def update( self, x: jnp.ndarray, t: jnp.ndarray) -> None:
        """Compute gradients and apply one optimizer step."""
        dW, dB = self._gradient(x, t)

        # Optimizer returns new immutable JAX arrays
        new_W, new_B = self.optimizer.update(self.weights, self.biases, dW, dB)

        self.weights = new_W
        self.biases  = new_B

        # Sync affine layer weight references so the next forward pass
        # uses the updated parameters
        for i, affine in enumerate(self.affine_layers):
            affine.W = self.weights[i]
            affine.B = self.biases[i]